Introduce steps

Baachi opened the pull request:
Hey guys!

The current library is really cool but i think it is not really scalebale. If somone needs to filter before all validators are called and an another need the oposite. This make it really hard to use this library.

My idea was to introduce steps. Now the library has a `StepInterface` which can modifiy/validate/whatever the item. Now can each developer decide, how his workflow looks like. Just instantiate a `StepInterface` and add them to the `Workflow` object.

I also added a `PriorityStepInterface`. This interface have a `getPriority` and with this method you can prioritize steps. Please note, that the developer can override the priority in the `addStep` method.

I think this will make the library more generic and easier to use.

> Please not, this is a __PoC__. Which means the code may not work and isn't finished yet.

Please provide Feedback!

YHuard commented on src/Ddeboer/DataImport/Workflow.php:
```diff
@@ -31,188 +25,47 @@ class Workflow
-        $this->getMappingItemConverter()->addMapping($fromField, $toField);
-
-        return $this;
+        $this->steps[$priority]->attach($step);
```
Small enhancement here, add ```return $this;``` thus you can chain your steps like this:

```php
$workflow
    ->addStep($myFirstStep)
    ->addStep($mySecondStep)
    ...
    ->process();
```

Baachi replied:
Agreed

YHuard commented:
Is the PriorityStepInterface a good idea? IMHO either all the steps should implement this interface or none of them. Otherwise it introduces some kind of confusion: if the user isn't aware that some classes have a priority when other haven't, he won't understand why some steps are run before other, especially when you chain the steps (like I commented above).

My opinion is you should forget the PriorityStepInterface and only set the priorities optionnaly when you call ```addStep```.

Also what about sharing variables between steps? I may use this functionality to re-use some code (e.g. data fetched in the database) in different steps. I see 3 different scopes for these variables: *step scope* (only available in the current step), *row scope* (available for all the steps in the current row) and *workflow scope* (available in every step). It would essentially be useful when using callbacks, I can't see other use cases

Baachi commented:
Thanks for your feedback:

> Is the PriorityStepInterface a good idea?

The idea behind the priorities is that some steps/filters have the requirement to be the first filter. But maybe we should document it and remove this interface.

> Also what about sharing variables between steps?

What do you mean? You can share variables/objects in the PHP way, use constructor injection or setter injection. Each project should have an (di) container or at least a registry so this shouldn't be a problem. If you have an use case, let me know. 

I don't like callbacks, it mades the code more complicated (welcome to the callback hell).

YHuard commented:
In my case, the CSV contains fields that have multiple values separated with commas and each value represent an object in the database. When I load the file, for each of this fields I have to: 
- explode the field ```explode(',', $field);``` (I do this in a value converter)
- apply converters on each value (```ucfirst``` or ```strtolower```)
- check if each value exists in the database: if it's true, I map the value with the entity, otherwise I throw a ValidationException

I have no other choice but to use callbacks to achieve that... 

Also I need to do "conditional validation" and that's when I need to share variables between steps. In a first step I check if a field exist in the database. If it exists, I ignore some columns of the CSV, otherwise I apply validators on them. In the callback writer I find myself reloading the entity that I already loaded in the first step. That's where I would use shared variables

Baachi commented:
@YHuard this should already possible. With this PR you can order your steps like you need it. Some users need to filter (like you) before the validation is called. This is currently not really possible.

Here are a little example, how it works:
```php
<?php

$workflow = (new Workflow())
   ->addStep(new FilterStep())
   ->addStep(new ValdationStep())
   ->addStep(new WriterStep())
   ->process()
```

And if you need dynamic validation, you can inject the `ValidationStep` into the `FilterStep` and set the constraints dynamicly.

YHuard commented:
Just out of curiosity, how would you inject the ```ValidationStep``` into the ```FilterStep```?

Instead of using shared variables I could also add new properties to the ```$item``` object therefore it isn't a real problem. My full workflow would the be the following:
- filter the data with ```OffsetFilter```
- apply basic ```ValidationFilter``` and ```ValueConverter``` 
- check existence in database and *store results in ```$item``` properties*
- apply a callback filter on some fields depending on the existence or not of the previous information
- and so on...

sagikazarmark commented:
This is definitely an interesting on, probably a fix for #99. However I am not sure it is a good idea to add writers as a step.

Writting logic should not be followed with any other steps, so writers are a very special case. I imagine a step as "something to do with an item". As mentioned in #99, ItemConverters and Filters could be completely replaced by Steps.

sagikazarmark commented on src/Ddeboer/DataImport/Step/ConverterStep.php:
```diff
@@ -0,0 +1,41 @@
+    public function clear()
+    {
+        $this->converters = new \SplObjectStorage();
+    }
```
What's the point of object storage at all? Isn't an array enough?

Baachi replied:
Because an `SplObjectStorage` consume less memory.

sagikazarmark commented:
After rereading the whole PR again, I think I finally understood the point of this PR. Your reasoning [here](https://github.com/ddeboer/data-import/pull/125#issuecomment-65955231) was the key to it (I missed that @ddeboer came up with the same as I did in the previous comment). However I am not sure that the scenario you mentioned should be in scope of this library. It makes things harder and I am not sure that cost worth it. But I have to admit that this is much more generic than the current workflow. So my concerns will probably go if I see it working.

sagikazarmark commented:
Further thinking about this: as I can see, you only introduced different steps for Filtering, Converting, etc to preserve the current structure of the existing components and to be able to have writers as steps. When introducing such a big BC break, this might not be necessary. Here is how I would do this:

1. Have a Step interface (done)
2. Have a generic StepCollection which accepts Steps
3. Rewrite all validators/filters/converters to be steps
4. Leave writers out of the step logic (as they should always be the last step)

Optionally the current idea from #99 of returning falsy value means filtering could be replaced with throwing an exception and catching it in the workflow. This way a Step can be forced to always return an actual item, any other "exceptional" cases can be handled with actual exceptions.

This would also mean that a step is "something that interacts with an item". In most of cases, a step would interact with an item. However, I admit that the logger scenario you mentioned does not fit into this. So I have an answer for this case as well: instead of having a StepCollection, you could have an ItemProcessingStep which accepts other steps or ItemProcessors which interact with the item. This way ItemConverters and Filters are merged and the Step logic you are introducing remains the same.

This became quite a long brain dump, so I finish this for now.

Baachi commented:
@sagikazarmark Thanks for your ideas! 

I also think about moving the writers to a step. I moved it just for consistence. I don't really like s specific handling. I think its easier to document that everything is a step. Currently its not easy enough to create an `Workflow` but we can introduce a `WorkflowBuilder` which made it easier to use.

I wouldn't throw exceptions if an element isn't valid. This will increase the memory usage and cpu cycles dramaticly. I know this, because twig throw an exception if a template could not be found and their ChainLoader, have a big problem with this. Remove the `try, catch` block increase the performance by 10%. I think we need a better solution for this.

sagikazarmark commented:
I didn't really think about the memory consumption of exceptions, but you are right. The ItemProcessingStep could still return a falsy value to indicate a filtering.

I admit that from a point of view, writting an item is just a step. But from another point of view it just does not fit into this workflow.

See this one:
``` php
<?php

$workflow = (new Workflow())
   ->addStep(new FilterStep())
   ->addStep(new ValdationStep())
   ->addStep(new WriterStep())
   ->addStep(new FilterStep())
   ->addStep(new ValdationStep())
   ->addStep(new WriterStep())
   ->process()
```

I don't think we should allow developer to do this. It should be in a different workflow. If we are saying that writer is the last step, it could be taken out of steps.

Or you could make the workflow immutable when a writer step is passed, but that is ugly.

sagikazarmark commented:
BTW memory consumption: would it cause too much memory usage if you clone the SplPriorityQueue before iterating over it?

Baachi commented:
Which `SplPriorityQueue` do you mean?

sagikazarmark commented on src/Ddeboer/DataImport/Workflow.php:
```diff
@@ -269,118 +104,45 @@ public function process()
-
-        return $converter;
-    }
+        ksort($tmp);
```
I mean this `SplPriorityQueue` which is an array now.

sagikazarmark replied:
So: would it cause too much memory usage if you clone the SplPriorityQueue before iterating over it?

Baachi commented:
I will work on this PR at the weekend.

sagikazarmark commented:
@Baachi ping, any update? Also, what about writers? I still think they should be separated from steps.

Baachi commented:
@sagikazarmark I made some commits, i hope the commit messages meaningful enough.
Writers are now a part of the `Workflow` class and i use the `SplPriorityQueue` instead of an simple array.

I also fixed the unit tests, so they should now pass. I also made some changes, so PHP 5.4 features are used, performance improvments and changes which simplyfied the code.

sagikazarmark commented:
@Baachi I looked through the changes. They seem to be cool. One thing to be discussed: the question of Filters and ItemConverters to be merged into ItemProcessors.

I could also imagine every converter to be renamed to steps, so you can create the following scenario (without messing with adding converter wrapper classes): filter out empty keys from the item, modify a value, do another item conversion.

Baachi commented:
@sagikazarmark Currently there is only one item converter (the `CallbackItemConverter`). I removed the `MappingItemConverter` and `NestedItemConverter` and introduced in commit  3cc72c3 an `MappingStep` which use the symfony property accessor component. 

In my opinion we can now drop the ItemConverter section completely.

sagikazarmark commented:
Awesome. The callback converter could be renamed to callback step. Then it
can probable be removed.

Baachi commented:
In my implementation the `CallbackItemConverter` does the same as the `CallbackFilter`, so i think the `CallbackItemConverter` is superfluous.

sagikazarmark commented:
If I understand well, the Filters and ItemConverters are kinda merged now.
Is that right?

Baachi commented:
Yes, ItemConverters and Filters are now merged.

sagikazarmark commented:
Great. Looking forward to it. Any todo before merging?

Baachi commented:
I must write some unit tests and update the documentation. And i would tag a new release before this will be merged because this PR will also bump the minimum php version to 5.4.

sagikazarmark commented:
Any schedule? Anything I can help?

Baachi commented:
I hope i will finish this PR this weekend. Sure any help is appreciated. Just send a PR to my branch and i will review and merge it. :)

sagikazarmark commented:
If the ItemConverters and Filters are the same now, shouldn't they be merged under a common name? Like ItemProcessor mentioned in the relevant PR

sagikazarmark commented:
Also, the order of filters is arbitrary, thus the getPriority should be removed.

sagikazarmark commented:
And probably a rebase would be needed.

sagikazarmark commented:
Any progress?

afurculita commented:
@Baachi any progress?

sagikazarmark commented:
@Baachi any progress?

ddeboer commented:
Heh take it easy, guys. :) I’ll try to review @Baachi’s most recent changes shortly.

gnat42 commented:
I've pushed a PR to @Baachi 's steps branch that merged master and fixed up a bunch of other items so that tests pass. There still are some 'hacks' if you will. One of the bigger issues is the use of the property accessor. @ddeboer I'd really like some feedback about this feature. I'm converting one project over to this branch since it requires the flexibility it provides.

Baachi commented:
So guys, thanks to @gnat42 we now have a new version including some bugfixes and new features:

  - I added the tests and they are green :fireworks: 
  - The `ValueConverterStep` and `ConverterStep` accept now any `callable`. The interfaces has been removed. Now you can easily use zend/filter or something else.
 - Please provide feedback. I think if anything is okay, i will merge it!

sagikazarmark commented:
I find the strict callable way for the mentioned Steps a bit weird. But it is only a minor thing. Looking at the whole picture, this is the best thing that happened this week so far.

Baachi commented:
@sagikazarmark Why do you think its weird? The monolog library does the same with his [processors](https://github.com/Seldaek/monolog/blob/master/src/Monolog/Processor/GitProcessor.php#L36). Its allows the developer to reuse filters easily.

Before:

```
$filter = (new FilterStep)
    ->add(new ZendFilterConverter(new \Zend\I18n\Filter\Alnum());

$workflow->add($filter);
$workflow->process();
```

After:
```
$filter = (new FilterStep)->add([new \Zend\I18n\Filter\Alnum(), 'filter']);
$workflow->add($filter)->process();
```

You can now use Zend\Filter and DMS\Filter without writing any additional class. I think this will be
a huge improvment for newbies.

sagikazarmark commented:
LOL, does it?

As I said, it is a minor thing. I am fine with that. From a developer point of view, it is harder to make sure a correct dependency is injected. From an API design point of view __invoke is a shot in PHP's head. (But I have to admit it eases a lot of things....for newbies mostly(?).)

Baachi commented:
I think so, but mabye its only my opinion. This is why i want to hear feedback, if everyone says its bullshit i can revert is. I agree with you, it isn't really a good design and it make it harder to debug for developers.

sagikazarmark commented:
There are many hobby programmers on the internet, but I think to provide good quality packages which last for a long time design is what really matters. Otherwise we end up like laravel: every version is a complete rewrite. I am generally against the solutions like this. But since the rest of the package depends on interfaces (I can even write my own steps) it is not a major issue. I would say, it makes the picture less beautiful, but not ugly yet. :stuck_out_tongue:

Baachi commented:
@sagikazarmark I generally agree with you, a clean design is a big point (this is why i won't like laravel). But personally i ofter write this code `->add(CallbackValueConverter(function() { /* some code */ }))` because it doesn't make sense to reuse this function (its project specify). Maybe i'm the only one which have 3 - 4 lines of code for each import/export.

sagikazarmark commented:
@Baachi I would say merge it. You can still revert it later.

sagikazarmark commented:
What's the situation here?

Baachi commented:
@ddeboer Are you okay with that?

sagikazarmark commented:
I absolutely am. One thing I am missing is some documentation in the README.

Also, I am not aware of any further plans, but this is a relatively big change, so before a new release I would wait for some time.

I am currently right before starting a project which will use this package extensively. So I can do some experiments with it.

Baachi commented:
@sagikazarmark I will improve the documentation in general, so i will add the new features.

Sure this is a big change, if this PR is merged, i will add a warning to the Readme sections, that the master branch isn't production ready.

ddeboer commented on src/Ddeboer/DataImport/Filter/DateTimeThresholdFilter.php:
```diff
@@ -47,14 +47,14 @@ public function __construct(
+    public function __invoke(array $item)
     {
         if ($this->threshold == null) {
             throw new \LogicException('Make sure you set a threshold');
```
Why isn’t the `$threshold` constructor argument required?

Baachi replied:
I don't know what was exactly the reason, i'm not the author of this. But you are right, it should be a constructor argument.

sagikazarmark commented:
Great, go ahead then.

ddeboer commented:
@Baachi Great work here!

Do you plan to work on the readme in this PR or in a separate one?

> i will add a warning to the Readme sections, that the master branch isn't production ready.

I don’t think such a warning is necessary. When we release this as 0.19.0, we should add a BC warning, however, telling users to stay at 0.18.0 if they’d like to continue using the library the old way.

ddeboer commented on src/Ddeboer/DataImport/Filter/DateTimeThresholdFilter.php:
```diff
@@ -47,14 +47,14 @@ public function __construct(
      * {@inheritDoc}
      */
-    public function filter(array $item)
+    public function __invoke(array $item)
```
Just checking: what exactly are the arguments for using `__invoke` over interfaces? Only that using external libraries (such as Zend) will be easier?

Baachi replied:
Not only this, in my opinion its easier to use for newbies and in my projects i do this `new CallbackFilter(function() { /* some code */ });`ever and ever again. So i thought its easier to use, but i know that this isn't a really good design decision. 

I already had a discussion with @sagikazarmark about that feature in this PR (see comments above). I with you have the same opinion as @sagikazarmark, i will revert them.

Baachi commented:
@ddeboer I can add a simple documentation in this pr (so you get a better overview what i've changed) and in the next few days/weeks i will provide a better PR. 

The complexitiy should now be lower than before, here is a quick snippet:

```php
$file = new \SplFileObject(date('y-m-d').'.xlsx', 'a+');

$workflow = (new Workflow(new ArrayReader($input))
    ->addWriter(new ExcelWriter($file))
    ->addStep($filters = new FilterStep())
    ->addStep($converts= new ValueConverterStep ())
;

$converts->add('created_at', new DateTimeValueConverter('y/m/d'));
$converts->add('updated_at', new DateTimeValueConverter('y/m/d'));
$converts->add('desc', 'strip_tags');
$converts->add('title', 'ucwords');

$filters->add(function($input)  { 
     return is_array($input) && empty($input['id']);
});

$workflow->process();
```

As you can see it isn't really hard to understand. For the end user, the change isn't really hard. Behind the scences it changed a lot, but this isn't visible for the end user.

This PR brings more flexibilty, because the user can now configure the workflow and the performance is better because the library now loads only the needed stuff.

Baachi commented:
@ddeboer please not this PR drops php 5.3 support.

gnat42 commented:
ooh ... I don't know if I'm ready for the drop of php 5.3. I need to double check. We have a couple of clients where their IT departments don't like using other non distro provided files.

sagikazarmark commented:
PHP 5.3 is not just EOL, but DEPRECATED. Dropping it should not be a
problem, especially not for corporate applications. My two cents only...

gnat42 commented:
I know. The truth is there are lots of systems out there running RHEL/Centos6 which comes with php 5.3 and receives backported security patches. All our new systems are newer php versions. Its really unfortunate that its such a pain for some LTS distros to update or provide 'side by side' versions if you will so that people can have the LTS distro but choose some of the quicker moving projects.

sagikazarmark commented:
For Ubuntu, there are update ppas.

Otherwise: PHPbrew or manual build

Baachi commented:
@gnat42 You can install both versions of php (5.3 and 5.5) on the same server. It shouldn't be a problem.

gnat42 commented:
So here's the thing. **I** know and use different versions of PHP on my server boxes. However some large organizations standardize their environments on a select few distributions and only install packages from the distribution only. I don't agree with that but I don't have power to change it. 

One thing that I wonder about as well. What is the benefit of dropping 5.3 support? Creating arrays with [ ] vs array() does it make a difference? Are there other features of 5.4 that are **so** great and make the library **so** much cleaner/faster/better/whatever that dropping 5.3 support is necessary? What does dropping it give us?

In any case I've double checked the project I'm using the data-import library with and they allow 5.4 so I'm fine with it. I just know of a few where they aren't as flexible.

sagikazarmark commented:
As a general answer to this: 5.x is unsafe, 5.x+1 is safer and faster. Also, 5.x which is DEPRECATED or EOL is a security threat. It does not get security updates, and new security issues are not getting fixed.
(In fact, some PHP 5.3 versions have very serious security issues.)

If it is not enough (actually, it is), in this particular case 5.4 gives us new and deprecated features, such as:

* new array notation
* traits
* callable typehint (which is extensively used in this PR)

For the full list: http://php.net/manual/en/migration54.php

Check reddit, this has been discussed a thousand (or more) times.

(My apologies, if I seem to be rude. I've answered this question several times)

Baachi commented:
@sagikazarmark Perfect answer.

sagikazarmark commented:
@Baachi thanks. Sometimes I don't have the patience to give a normal answer for questions like this.

Where are we now with this PR? Anything that should be done before merging? Or is it in review state by @ddeboer?

Baachi commented:
I would wait for @ddeboer answer and then i merge it and then the i work on the documentation.

ddeboer commented:
:+1: for requiring PHP 5.4

@Baachi If it’s not too much of a hassle: would it be possible to squash commits?

Baachi commented:
@ddeboer I squashed my commits!

sagikazarmark commented:
Looking forward to merge

ddeboer commented:
:fireworks: Great work, @Baachi and others that helped out!

sagikazarmark commented:
Yeah, @Baachi respect.

Baachi commented:
:fireworks: Thank you
